### PR TITLE
less copying, primitive builder

### PR DESCRIPTION
**Pull request by scottf**

### CODE REVIEW

File: src/main/java/io/nats/client/impl/Headers.java [hunk 1]
```diff
@@ -390,6 +394,40 @@ public ByteArrayBuilder appendSerialized(ByteArrayBuilder bab) {
...
+				bytes = value.getBytes(US_ASCII);
+				System.arraycopy(bytes, 0, dest, destPosition, bytes.length);
+				destPosition += bytes.length;
+
+				dest[destPosition++] = CR;
+				dest[destPosition++] = LF;
+			}
+		}
+		dest[destPosition++] = CR;
+		dest[destPosition] = LF;
```
Comment (MauriceVanVeen) [inline src/main/java/io/nats/client/impl/Headers.java:426]: is the `destPosition++` missing here on purpose?

File: src/main/java/io/nats/client/impl/NatsMessage.java [hunk 2]
```diff
@@ -129,28 +130,29 @@ protected boolean calculateIfDirty() {
...
             // reply to if it's there
             if (replyToLen > 0) {
-                bab.append(replyTo).append(SP);
+                bab.append(replyTo.getBytes(UTF_8)).append(SP);
             }
 
             // header length if there are headers
             if (hdrLen > 0) {
-                bab.append(Integer.toString(hdrLen)).append(SP);
+                bab.append(Integer.toString(hdrLen).getBytes(US_ASCII)).append(SP);
```
Comment (MauriceVanVeen) [inline src/main/java/io/nats/client/impl/NatsMessage.java:149]: the `subject` and `replyTo` use `UTF_8`, while the header and payload use `US_ASCII`
not quite sure, but shouldn't it be either utf8 or ascii?

File: src/main/java/io/nats/client/impl/Headers.java [hunk 3]
```diff
@@ -390,6 +394,40 @@ public ByteArrayBuilder appendSerialized(ByteArrayBuilder bab) {
...
+
+				bytes = value.getBytes(US_ASCII);
+				System.arraycopy(bytes, 0, dest, destPosition, bytes.length);
+				destPosition += bytes.length;
+
+				dest[destPosition++] = CR;
+				dest[destPosition++] = LF;
+			}
+		}
+		dest[destPosition++] = CR;
```
Comment (MauriceVanVeen) [inline src/main/java/io/nats/client/impl/Headers.java:425]: I'm wondering if these CRLF can/should be removed.

Looking in the [`NatsConnectionWriter`](https://github.com/nats-io/nats.java/pull/776/files#diff-7b435bb1d33408d9aa01983b0c43086b0da7f479636890d2f05d248b0971f676L139), it looks like the extra CRLF was not present before?

File: src/main/java/io/nats/client/support/BuilderBase.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    private int _defaultCharsetAllocationSize() {
+        return defaultCharset == US_ASCII ? DEFAULT_ASCII_ALLOCATION : DEFAULT_OTHER_ALLOCATION;
+    }
+
+    public static int bufferAllocSize(int atLeast, int blockSize) {
+        return atLeast < blockSize
+            ? blockSize
+            : ((atLeast + blockSize) / blockSize) * blockSize;
```
[the remark this answers is not in the document]
Comment (scottf) [reply src/main/java/io/nats/client/support/BuilderBase.java:119]: I want to allocate on blocksize multiples, planning for multiples of 32 which is supposedly what Java likes to allocate on 64 bit processors. Also this gives 64 bytes when the atLeast is 32. It's arbitrary, I probably didn't need to over think it.

File: src/main/java/io/nats/client/impl/Headers.java (hunk 1 above)
Comment (scottf) [reply src/main/java/io/nats/client/impl/Headers.java:426]: Yeah. It's the last usage, IntelliJ reminded me that destPosition isn't used any more. I'm returning the length of the data, not the position.

File: src/main/java/io/nats/client/support/BuilderBase.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (MauriceVanVeen) [reply src/main/java/io/nats/client/support/BuilderBase.java:119]: That makes sense. Don't know what would be better, but this way sounds good.

File: src/main/java/io/nats/client/impl/NatsMessage.java (hunk 2 above)
Comment (scottf) [reply src/main/java/io/nats/client/impl/NatsMessage.java:149]: Numbers will never be anything other than ascii, so I don't have to worry that a character in the string of numbers could be utf8. Here is an example protocol line.  
```
HPUB SUBJECT REPLY 48 55␍␊NATS/1.0␍␊Header1: X␍␊Header1: Y␍␊Header2: Z␍␊␍␊PAYLOAD␍␊
```

This line of code could have been left alone, but the old version called append with a string, which checked what the default encoding of the bab and got those bytes, then made another call to the append that takes bytes. I just reduced the amount of code being executed by doing more manually.

Comment (MauriceVanVeen) [reply src/main/java/io/nats/client/impl/NatsMessage.java:149]: Ah, didn't recognise them just being numbers, thanks :)

**Review by MauriceVanVeen — APPROVED** [review]

**Pull request merged**